### DISCUSSION TITLE
Make `App::run` non-blocking on web

### DISCUSSION THREAD

**Discussion by Amex1048** [bevyengine/bevy#11258] [Q&A] [answered]
Hi! I managed to export my rust function to `js` via `wasm_bindgen`, but the question is if the function call would be non-blocking. Docs state, that I can make `App::run` non-blocking by setting `WinitSettings::return_from_run` flag, but it is not supported on web. Is there any other way to make my `js` API non-blocking?

**Comment by daxpedda** [comment, marked as answer]
You can't really block on Web.

The way it currently works is that `App::run()` (which calls [`winit::event_loop::EventLoop::run()`](https://docs.rs/winit/0.29.9/winit/event_loop/struct.EventLoop.html#method.run)) will throw (as in JS [`throw`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw)) which completely exits Rust. The rest of the application will continue working because it's callback based, e.g. the browser will call back into Rust when things happen.

So I'm not entirely sure what it is you are trying to do, but calling your Rust function which calls `App::run()` from JS won't block, but it will throw.

**Reply by Amex1048** [reply]
Thanks!

**Reply by maximetinu** [reply]
That is what [this Winit exception](https://github.com/rust-windowing/winit/blob/master/src/platform_impl/web/event_loop/mod.rs#L72), `Using exceptions for control flow, don't mind me. This isn't actually an error!`, does
